go/oasis-node/cmd/debug/txsource/workload/queries.go

```diff
@@ -0,0 +1,455 @@
+	})
+	if err != nil {
+		return fmt.Errorf("runtimeClient.GetBlock, round: latest: %w", err)
+	}
+	latestRound := latestBlock.Header.Round
+
+	// Select round at which queries should be done. Earliest (round=1)
+	// is special cased with increased probability.
+	var round uint64
+	p := rng.Float32()
+	switch {
+	case p < queriesEarliestHeightRatio:
+		round = 1
+	default:
+		// [1, latestRound]
+		round = uint64(rng.Int63n(int64(latestRound) + 1))
+	}
+
+	// GetBlock.
+	block, err := q.runtime.GetBlock(ctx, &runtimeClient.GetBlockRequest{
+		RuntimeID: q.runtimeID,
+		Round:     round,
+	})
+	if err != nil {
+		return fmt.Errorf("runtimeClient.GetBlock, round: %d: %w", round, err)
+	}
+	block2, err := q.runtime.GetBlockByHash(ctx, &runtimeClient.GetBlockByHashRequest{
+		RuntimeID: q.runtimeID,
+		BlockHash: block.Header.EncodedHash(),
+	})
+	if err != nil {
+		return fmt.Errorf("runtimeClient.GetBlockByHash, hash: %s: %w", block.Header.EncodedHash(), err)
+	}
+	if block.Header.EncodedHash() != block2.Header.EncodedHash() {
+		return fmt.Errorf("Expected equal blocks, got: byRound: %s byHash: %s", block.Header.EncodedHash(), block2.Header.EncodedHash())
+	}
+
+	_, err = q.runtime.QueryTxs(ctx, &runtimeClient.QueryTxsRequest{
+		RuntimeID: q.runtimeID,
+		Query: runtimeClient.Query{
+			RoundMin: 0,
+			RoundMax: round,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("runtimeClient.QueryTxs: %w", err)
+	}
+
+	q.logger.Debug("Done runtime queries",
+		"latest_round", latestRound,
+		"block", block,
+		"round", round,
+	)
+
+	return nil
+}
+
+func (q *queries) Run(gracefulExit context.Context, rng *rand.Rand, conn *grpc.ClientConn, cnsc consensus.ClientBackend, fundingAccount signature.Signer) error {
+	ctx := context.Background()
+
+	q.logger = logging.GetLogger("cmd/txsource/workload/queries")
+
+	q.consensus = cnsc
+	q.registry = registry.NewRegistryClient(conn)
+	q.runtime = runtimeClient.NewRuntimeClient(conn)
+	q.scheduler = scheduler.NewSchedulerClient(conn)
+	q.staking = staking.NewStakingClient(conn)
+
+	// TODO: could add a methods to get consensus parameters directly.
+	doc, err := q.consensus.StateToGenesis(ctx, consensus.HeightLatest)
+	if err != nil {
+		return fmt.Errorf("consensus.StateToGenesis error: %w", err)
+	}
+	q.epochtimeParams = doc.EpochTime.Parameters
+	q.stakingParams = doc.Staking.Parameters
+	q.schedulerParams = doc.Scheduler.Parameters
+
+	// Setup simple-keyvalue runtime info.
+	err = q.runtimeID.UnmarshalHex(viper.GetString(CfgRuntimeID))
+	if err != nil {
+		q.logger.Error("runtime unmarshal error",
+			"err", err,
+			"runtime_id", viper.GetString(CfgRuntimeID),
+		)
+		return fmt.Errorf("Runtime unmarshal: %w", err)
+	}
+
+	for {
+		block, err := q.consensus.GetBlock(ctx, consensus.HeightLatest)
+		if err != nil {
+			return fmt.Errorf("consensus.GetBlock error: %w", err)
+		}
+
+		// Select height at which queries should be done. Earliest (height=1)
+		// is special cased with increased probability.
```

Comment:
why extra likely on the earliest?

should we also make it extra likely to ask for the latest?

Reply:
We could (however not `consensus.HeightLatest` without special casing that case in some of the checks). I actually had that implemented as well but then figured that most of the latest queries are already implicitly tested by other workloads and sanity checks, and that this workload should instead focus on querying older data (as it was mentioned in the initial idea for it: https://github.com/oasislabs/oasis-core/issues/2506#issuecomment-594524062).

With that in mind, special casing the height 1 still does make some sense i think.

Reply:
ok